syntax = "proto3";

package roleplay_bot.ai;

service StoryWriter {
  rpc InitializeSession (InitializeSessionRequest) returns (InitializeSessionResponse);
  rpc GetSessionConfig (GetSessionConfigRequest) returns (GetSessionConfigResponse);
  rpc WriteStory (WriteStoryRequest) returns (WriteStoryResponse);
  rpc GetStoryHistory (GetStoryHistoryRequest) returns (GetStoryHistoryResponse);
  rpc DeleteSession (DeleteSessionRequest) returns (DeleteSessionResponse);
}

enum Model {
  Gemini_1_5_Pro = 0;
}

message InitializeSessionRequest {
  Model model = 1;
  string system_config = 2;
  double temperature = 3;
  double top_p = 4;
  uint32 top_k = 5;
  uint32 max_length = 6;
}

message InitializeSessionResponse {
  string session_id = 1;
}

message WriteStoryRequest {
  string session_id = 1;
  string prompt = 2;
}

message WriteStoryResponse {
  string story = 1;
}

message GetStoryHistoryRequest {
  string session_id = 1;
}

message StoryDetails {
  string prompt = 1;
  string story = 2;
  uint64 created_time = 3;
}

message GetStoryHistoryResponse {
  string system_config = 1;
  repeated StoryDetails story = 2;
}

message GetSessionConfigRequest {
  string session_id = 1;
}

message GetSessionConfigResponse {
  Model model = 1;
  string system_config = 2;
  double temperature = 3;
  double top_p = 4;
  uint32 top_k = 5;
  uint32 max_length = 6;
  uint64 created_time = 7;
}

message DeleteSessionRequest {
  string session_id = 1;
}

message DeleteSessionResponse {
  string status_message = 1;
}
